### DISCUSSION TITLE
Create ZStream.fromHub(...)

### DISCUSSION THREAD

**Discussion by abiketov** [zio/zio#8488] [Q&A]
I'm having issue with creating ZStream from Hub to receive messages that are published into the hub. 
If I use example from the documentation with ZStream.fromHubScoped everything works
```
for {
  promise <- Promise.make[Nothing, Unit]
  hub     <- Hub.bounded[String](2)
  scoped  = ZStream.fromHubScoped(hub).tap(_ => promise.succeed(()))
  stream   = ZStream.unwrapScoped(scoped)
  fiber   <- stream.take(2).runCollect.fork
  _       <- promise.await
  _       <- hub.publish("Hello")
  _       <- hub.publish("World")
  _       <- fiber.join
} yield ()
```
But if I try to use ZStream.fromHub() like below, basically same code:
```
for {
    promise <- Promise.make[Nothing, Unit]
    hub     <- Hub.bounded[String](2)
    stream = ZStream.fromHub(hub).tap(_ => promise.succeed(()))
    fiber     <- stream.take(2).runCollect.fork
    _         <- promise.await
    _         <- hub.publish("Hello")
    _         <- hub.publish("World")
    collected <- fiber.join
    output    <- console
    _         <- ZIO.foreach(collected)(output.printLine(_)).orDie
  } yield ()
```
after start it just hangs.

**Comment by adamgfraser** [comment]
They are not the same. The signature of `ZSteam.fromHubScoped` is:

```scala
def fromHubScoped[A](hub: => Hub[A]): ZIO[Scope, Nothing, ZStream[Any, Nothing, A]]
```

Notice that it is an effect returning a stream. The documentation states:

```
The scoped effect describes subscribing to receive messages from the hub while the stream describes taking messages from the hub.
```

So in the original example you are subscribing to the hub, then once you have subscribed to the hub you are completing the `promise` to signal to the publishers that they may begin publishing. You are thus guaranteed to see both of the values they publish and the stream completes after taking two elements.

In contrast, in your example there is no guarantee that you have subscribed to the hub by the time you complete the promise. As a result, the publishers may publish their values before you are subscribed and you never see them, so you never receive two elements and suspend indefinitely.
